Tests: Refactor tests to use only Jest API

gziolo opened the pull request:
This is a follow up for #1382. This PR performs all remaining refactorings to get rid of Mocha specific code:
- Sinon test doubles
- Sinon fake timers
- Chai assertions
- Chai lifecycle methods

This change makes test execution a bit faster.

#### Before

> real **0m5.961s**
> user 0m30.956s
> sys 0m2.355s

#### After
> real    **0m4.529s**
> user    0m22.167s
> sys     0m1.724s


### Testing
Make sure `npm test` passes.

gziolo committed:
Tests: Update tests to use Jest API
It was as easy as executing jest-codemods and eslint --fix space-in-parens.

gziolo committed:
Tests: Refactor tests to use Jest test doubles instead of Sinon

gziolo commented:
@nb This concludes Mocha to Jest migration for Gutenberg! This PR gives an overview how much code would be updated in the context of Automattic/wp-calypso. Unforntunately sinon changes aren't covered with `jest-codemods`. Maybe there are other codemods, because it was quite easy to change manually. Most of the changes were applied with the combination:
```bash
# jest-codemods <file-path>
# npm run lint -- --fix space-in-parens

nylen commented on blocks/api/test/serializer.js:
```diff
@@ -1,11 +1,6 @@
-/**
- * WordPress dependencies
- */
 import { createElement, Component } from 'element';
```
This section should now be titled `WordPress dependencies`, not `External dependencies`.

gziolo replied:
I see. `jest-codemods` wasn't smart enough to use those conventions :)

nylen commented on blocks/library/freeform/test/naiveCss2Jsx.js:
```diff
@@ -1,10 +1,6 @@
-/**
- * Internal dependencies
- */
 import { naiveCss2Jsx } from '../format-list';
```
This section should now be titled `Internal dependencies`, not `External dependencies`.

nylen commented on editor/block-mover/test/mover-label.js:
```diff
@@ -1,11 +1,6 @@
-/**
- * Internal dependencies
- */
 import { getBlockMoverLabel, getMultiBlockMoverLabel } from '../mover-label';
```
This section should now be titled `Internal dependencies`, not `External dependencies`.

nylen commented on editor/test/actions.js:
```diff
@@ -1,11 +1,6 @@
-
-/**
- * Internal dependencies
- */
```
This section should now be titled `Internal dependencies`, not `External dependencies`.

nylen commented on editor/utils/test/dom.js:
```diff
@@ -1,35 +1,30 @@
-/**
- * Internal dependencies
- */
 import { isEditableElement } from '../dom';
```
This section should now be titled `Internal dependencies`, not `External dependencies`.

nylen commented on editor/utils/test/undoable-reducer.js:
```diff
@@ -1,11 +1,6 @@
-/**
- * Internal dependencies
- */
 import { undoable, combineUndoableReducers } from '../undoable-reducer';
```
This section should now be titled `Internal dependencies`, not `External dependencies`.

nylen commented on editor/utils/test/url.js:
```diff
@@ -1,25 +1,20 @@
-/**
- * Internal dependencies
- */
 import { addQueryArgs } from '../url';
```
This section should now be titled `Internal dependencies`, not `External dependencies`.

nylen commented on element/test/index.js:
```diff
@@ -1,86 +1,81 @@
-/**
- * Internal dependencies
- */
 import { createElement, renderToString, concatChildren, switchChildrenNodeName } from '../';
```
This section should now be titled `Internal dependencies`, not `External dependencies`.

gziolo replied:
I'm wondering if it would be possible to add Eslint rule that would verify such use case.

nylen commented on CONTRIBUTING.md:
```diff
@@ -38,7 +38,7 @@ Gutenberg contains both PHP and JavaScript code, and encourages testing and code
 ### JavaScript Testing
 
-Tests for JavaScript use [Jest](http://facebook.github.io/jest//) as the test runner. If needed, you can also use [Sinon](http://sinonjs.org/) for mocking and [Enzyme](https://github.com/airbnb/enzyme) for React component testing.
+Tests for JavaScript use [Jest](http://facebook.github.io/jest//) as the test runner. If needed, you can also use [Enzyme](https://github.com/airbnb/enzyme) for React component testing.
```
Super minor: Double slash in this URL.

nylen commented on blocks/test/full-content.js:
```diff
@@ -124,10 +123,7 @@ describe( 'full post content fixture', () => {
-				parserOutputExpected,
-				format( 'File \'%s.parsed.json\' does not match expected value', f )
-			);
+			).toEqual( parserOutputExpected );
```
When these tests fail, debugging them is pretty tricky without these messages in place.  Let's preserve them (this one and the 2 below it), even if we have to do it via a `try / catch` or some other non-pretty mechanism.

Context: https://github.com/facebook/jest/issues/1965

gziolo replied:
This is what I see when I make JSON.parse throw an error:

> ● full post content fixture › core__html
> 
>     SyntaxError: Unexpected token x in JSON at position 0
>         at Object.parse (native)
>       
>       at Object.<anonymous> (blocks/test/full-content.js:123:36)
>       at Promise.resolve.then.el (node_modules/p-map/index.js:42:16)
>       at process._tickCallback (internal/process/next_tick.js:109:7)

It seems okey from my perspective. We can also code our own matcher which would print error using [expect.extend](https://facebook.github.io/jest/docs/en/expect.html#expectextendmatchers).

nylen replied:
A JSON parse error is not a realistic failure mode here.  When these tests fail, we have a valid JSON object but a mismatch in its content.  Or a mismatch in the serialized string content.  It's difficult to tell the difference between these two, so we need to keep some code that prints the filename.

I see that this is done in the latest version of the PR. :+1:

nylen commented on components/form-token-field/test/index.js:
```diff
@@ -126,147 +124,147 @@ maybeDescribe( 'FormTokenField', function() {
+			expect( getSuggestionsText() ).toEqual( fixtures.matchingSuggestions.at );
 		} );
 
-		it( 'should manage the selected suggestion based on both keyboard and mouse events', test( function() {
```
What was this `test()` wrapper?

gziolo replied:
Sinon exposes `test` method to enable fake timers. We have fake timers enabled by default in the current Jest configuration.

nylen commented on editor/block-mover/test/mover-label.js:
```diff
@@ -23,9 +18,9 @@ describe( 'block mover', () => {
-			);
+			) ).toBe(
+                `Block "${ type }" is at the beginning of the content and can’t be moved up`
+            );
```
The new indentation here looks wrong.  In fact, I would expect this to cause a lint error and we should investigate why it isn't.

ntwb replied:
Both lines 22 & 23 here are indented with spaces and not tabs

As to why ESLint is missing that I've no idea /shrug

nylen commented on editor/test/state.js:
```diff
@@ -1033,15 +1032,15 @@ describe( 'state', () => {
 			const state = store.getState();
 
-			expect( Object.keys( state ) ).to.have.members( [
+			expect( Object.keys( state ) ).toEqual( expect.arrayContaining( [
```
Not a critique for this PR, but I'm not a huge fan of this syntax.

gziolo replied:
I would blame `jest-codemod` here. I'm sure we can refactor later using [expect.objectContaining ]https://facebook.github.io/jest/docs/en/expect.html#expectobjectcontainingobject).

nylen replied:
The `toEqual( expect.someWeirdObject` syntax in general is strange.  Again not something we can really change here.

nylen commented on package.json:
```diff
@@ -77,9 +75,6 @@
     "sass-loader": "^6.0.3",
-    "sinon": "^2.1.0",
-    "sinon-chai": "^2.9.0",
-    "sinon-test": "^1.0.2",
```
:tada: :tada: :tada:

nylen commented on package.json:
```diff
@@ -51,11 +51,9 @@
     "babel-plugin-transform-runtime": "^6.23.0",
     "babel-preset-env": "^1.4.0",
     "babel-traverse": "^6.24.1",
-    "chai": "^3.5.0",
```
:tada:

nylen commented on package.json:
```diff
@@ -51,11 +51,9 @@
     "concurrently": "^3.4.0",
     "cross-env": "^3.2.4",
     "deep-freeze": "0.0.1",
-    "dirty-chai": "^1.2.2",
```
:tada:

nylen reviewed (changes requested):
Some minor revisions needed but overall I'm a fan.  Can we now also remove `mocha: true` from the eslint config?

nb commented:
Great work, @gziolo. The `sinon` changes don't worry me much.

I have a feeling we will get the most benefits from `jest` if we start using snapshots for component/integration tests, any chance we can try this in Gutenberg? Maintaining all of those fixtures… 💣

gziolo committed:
Tests: Fix all issues introduced when executing jest-codemods

gziolo committed:
Tests: Wrap expectations in blocks tests with try / catch to print more detailed info

gziolo commented:
@nylen Thanks for catching all issues introduced with `jest-codemods` run. I fixed all of them. I also added `try` / `catch` clause to make sure we display more user friendly message with removed debug like:
> format( 'File \'%s.parsed.json\' does not match expected value', f )

Let me know if that works for you until Jest adds a better way to do it.

nb commented:
> > I guess the best way would be to refactor this code to use snapshots instead. We can do it in the follow up PR.

What’s your thinking about what tests should use snapshots? A lot of the tests touched in this PR should better stay the way they are. Converting component and fixture tests to snapshots would be awesome, though. Agreed that a separate PR is better.

nylen committed:
Preserve original error details when re-throwing

nylen commented:
> Let me know if that works for you until Jest adds a better way to do it.

In 4e481fe I've also preserved the original error message which contains the actual+expected values and the differences between them.

Thanks for working on this; it's looking good to me.

gziolo commented:
> In 4e481fe I've also preserved the original error message which contains the actual+expected values and the differences between them.

Yes, it makes perfect sense. 👍 

> What’s your thinking about what tests should use snapshots? A lot of the tests touched in this PR should better stay the way they are. Converting component and fixture tests to snapshots would be awesome, though. Agreed that a separate PR is better.

@nb I rather consider snapshot testing as complementary. When you develop component, you should also use TDD techniques and introduce snapshot based tests once you have something working to make sure it doesn't change later. Definitely tests where you create tons of fixtures are perfect match to use snapshots, too.

nb commented:
> I rather consider snapshot testing as complementary. 

This sounds like too many tests  – why can't we use snapshots to TDD components? We should start with some small output and then build up.

gziolo commented:
> This sounds like too many tests – why can't we use snapshots to TDD components? We should start with some small output and then build up.

Let's see how it goes. I don't have any experience with snapshots so I'm only blindly predicting :D
I opened #1841 which is a perfect use case for snapshots!
